### PR TITLE
cilium-kube-proxy ipv6-only implementation

### PR DESCRIPTION
**Pull request by aanm**
Waiting on #131

### CODE REVIEW

File: contrib/vagrant/start.sh [hunk 1]
```diff
@@ -84,17 +84,21 @@ function write_footer() {
     ipv6_addr="${2}"
     filename="${3}"
 
-    if [ -n "${IPV4}" ]; then
+    if [ -n "${IPV4}" ] && [ "${IPV4}" -ne "0" ]; then
         ipv4_options="--ipv4 --ipv4-range 10.${index}.0.1 "
     fi
 
+    if [ -n "${K8S}" ]; then
```
Comment (tgraf) [inline contrib/vagrant/start.sh:91]: Does this need `K8S=0` handling as well?

File: daemon/daemon/k8s_watcher.go [hunk 2]
```diff
@@ -279,36 +309,62 @@ func (d *Daemon) syncLB() {
...
 			isServerPresent := false
 
-			serverIndex := 0
+			serverIndex := uint16(0)
 
 			if len(se.IPs) != 0 {
-				log.Debugf("# cilium lb update-service %s %d %d %d %d %s %d",
-					svcInfo.IP, svcPort.Port, serverIndex, len(svcInfo.Ports), svcPort.ServiceID,
-					"::", 0)
+				svc6k := lbmap.NewService6Key(svcInfo.IP, svcPort.Port, serverIndex)
```
Comment (tgraf) [inline daemon/daemon/k8s_watcher.go:320]: Can you add a check whether the IP is IPv6? (`To4() == nil`)

File: doc/k8s.md [hunk 3]
```diff
@@ -1 +1,40 @@
...
+# Cilium Loadbalancer and Kubernetes
+
+Cilium provides a loadbalancer by its own which can be used as a replacement to kube-proxy.
+
+This feature is still under development but this readme file will be kept updated as
+possible.
+
+## Kubernetes loadbalancer (kube-proxy)
+
+Kubernetes has develop the Services abstration which provides the user the ability to
```
Comment (tgraf) [inline doc/k8s.md]: developed

File: doc/k8s.md [hunk 4]
```diff
@@ -1 +1,40 @@
...
+load balance network traffic to different pods. This abstraction allows the pods
+reaching out to other pods by a single IP address, a virtual IP address, without
+knowing all the pods that are running that particular service.
+
+Kube-proxy, installed on every node, watches for endpoints and services addition and
+removal on the kube-master which allows him to apply the necessary enforcement on
```
Comment (tgraf) [inline doc/k8s.md]: allows it to

File: doc/k8s.md [hunk 5]
```diff
@@ -1 +1,40 @@
...
+
+## What do I need to change in kubernetes?
+
+If you are using the kubernetes `hack/local-up-cluster.sh` script you only need to
+apply the cilium kubernetes [patch](../examples/kubernetes/kubernetes-v1.3.7.patch) and
+use the same environment variables that we use [here](../examples/kubernetes/env-kube.sh)
+
+Otherwise you need to disable the local kube-proxy and start the kubernetes apiserver
+with the `--service-cluster-ip-range="f00d:1::/112"`. You can choose your own IPv6
+prefix, but keep the mask to `/112` or bigger (>=112).
```
Comment (tgraf) [inline doc/k8s.md:35]: with bigger you mean bigger IP space?

**Review by tgraf — CHANGES_REQUESTED** [review]

File: contrib/vagrant/start.sh (hunk 1 above)
Comment (aanm) [reply contrib/vagrant/start.sh:91]: No. For example I'm currently starting the script as:

```
IPV4=0 RELOAD=1 K8S=1 NFS=1 ./start.sh
```

where I don't want IPv4 but I want K8S

Comment (tgraf) [reply contrib/vagrant/start.sh:91]: Right but `IPV4=0 K8S=0 ./start.sh` would not work as expected

File: daemon/daemon/k8s_watcher.go (hunk 2 above)
Comment (aanm) [reply daemon/daemon/k8s_watcher.go:320]: Yes, I will change this a bit, probably on a new commit. I'll detect if the IP is IPv4 or IPv6 and add it on the respective service map

File: doc/k8s.md (hunk 5 above)
Comment (aanm) [reply doc/k8s.md:35]: No, bigger mask

File: contrib/vagrant/start.sh (hunk 1 above)
Comment (aanm) [reply contrib/vagrant/start.sh:91]: True but `IPV4` is set to `1` but default. Should I change `if [ -n "${K8S}" ]; then` to `if [ -n "${K8S}" ] && [ "${K8S}" -ne "0" ]; then`? and the `NFS` as well?

Comment (tgraf) [reply contrib/vagrant/start.sh:91]: I think we should change all of them. Can be done outside of this PR though.

File: daemon/daemon/k8s_watcher.go (hunk 2 above)
Comment (tgraf) [reply daemon/daemon/k8s_watcher.go:320]: Fine with me. Please mark the PR as IPv6 only for now then.

File: doc/k8s.md (hunk 5 above)
Comment (tgraf) [reply doc/k8s.md:35]: OK

File: contrib/vagrant/start.sh (hunk 1 above)
Comment (aanm) [reply contrib/vagrant/start.sh:91]: https://github.com/cilium/cilium/projects/2#card-203378

File: daemon/daemon/k8s_watcher.go (hunk 2 above)
Comment (aanm) [reply daemon/daemon/k8s_watcher.go:320]: https://github.com/cilium/cilium/projects/2#card-203385

**Comment by aanm** [comment]
@tgraf all comments addressed

**Review by tgraf — APPROVED** [review]

**Pull request merged**